hudi-client/src/test/java/org/apache/hudi/HoodieClientTestHarness.java

```diff
@@ -0,0 +1,206 @@
+    sqlContext = new SQLContext(jsc);
+  }
+
+  protected void initSparkContexts() {
+    initSparkContexts("TestHoodieClient");
+  }
+
+  protected void cleanupSparkContexts() {
+    if (sqlContext != null) {
+      logger.info("Clearing sql context cache of spark-session used in previous test-case");
+      sqlContext.clearCache();
+      sqlContext = null;
+    }
+
+    if (jsc != null) {
+      logger.info("Closing spark context used in previous test-case");
+      jsc.close();
+      jsc.stop();
+      jsc = null;
+    }
+  }
+
+  protected void initTempFolderAndPath() throws IOException {
+    folder = new TemporaryFolder();
+    folder.create();
+    basePath = folder.getRoot().getAbsolutePath();
+  }
+
+  protected void cleanupTempFolderAndPath() throws IOException {
+    if (basePath != null) {
+      new File(basePath).delete();
+    }
+
+    if (folder != null) {
+      logger.info("Explicitly removing workspace used in previously run test-case");
+      folder.delete();
+    }
+  }
+
+  protected void initFileSystem() {
+    if (basePath == null) {
+      throw new IllegalStateException("The base path has not been initialized.");
+    }
+
+    if (jsc == null) {
+      throw new IllegalStateException("The Spark context has not been initialized.");
+    }
+
+    fs = FSUtils.getFs(basePath, jsc.hadoopConfiguration());
+    if (fs instanceof LocalFileSystem) {
+      LocalFileSystem lfs = (LocalFileSystem) fs;
+      // With LocalFileSystem, with checksum disabled, fs.open() returns an inputStream which is FSInputStream
+      // This causes ClassCastExceptions in LogRecordScanner (and potentially other places) calling fs.open
+      // So, for the tests, we enforce checksum verification to circumvent the problem
+      lfs.setVerifyChecksum(true);
+    }
+  }
+
+  protected void initFileSystemWithDefaultConfiguration() {
+    fs = FSUtils.getFs(basePath, new Configuration());
+    if (fs instanceof LocalFileSystem) {
+      LocalFileSystem lfs = (LocalFileSystem) fs;
+      // With LocalFileSystem, with checksum disabled, fs.open() returns an inputStream which is FSInputStream
+      // This causes ClassCastExceptions in LogRecordScanner (and potentially other places) calling fs.open
+      // So, for the tests, we enforce checksum verification to circumvent the problem
+      lfs.setVerifyChecksum(true);
+    }
+  }
+
+  protected void cleanupFileSystem() throws IOException {
+    if (fs != null) {
+      logger.warn("Closing file-system instance used in previous test-run");
+      fs.close();
+    }
+  }
+
+  protected void initTableType() throws IOException {
+    if (basePath == null) {
+      throw new IllegalStateException("The base path has not been initialized.");
+    }
+
+    if (jsc == null) {
+      throw new IllegalStateException("The Spark context has not been initialized.");
+    }
+
+    HoodieTestUtils.initTableType(jsc.hadoopConfiguration(), basePath, getTableType());
+  }
+
+  protected void cleanupTableType() {
```

Comment:
should this be abstract?  docs on what the intended overridden behavior would be nice

Reply:
Currently, We did not get an instance of `HoodieTableMetaClient` or init a class-level instance field in `initTableType` method. So we may need not to mark this method as `abstract` currently. I will try to refine this after this PR soon.